File: day4/golang/day4.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	stage1()
	stage2()
}
func stage1() {
	file, _ := os.ReadFile("../input.txt")
	out := 0
	for _, i := range strings.Split(string(file), "\n") {
		var hold []int
		for _, j := range strings.Split(i, ",") {
			val := strings.Split(j, "-")
			first, _ := strconv.Atoi(val[0])
			last, _ := strconv.Atoi(val[1])
			hold = append(hold, first)
			hold = append(hold, last)
		}
		if fullyHold(hold) {
			out += 1
		}
	}
	fmt.Println("stage 1 =", out)
}
func stage2() {
	file, _ := os.ReadFile("../input.txt")
	out := 0
	for _, i := range strings.Split(string(file), "\n") {
		var hold []int
		for _, j := range strings.Split(i, ",") {
			val := strings.Split(j, "-")
			first, _ := strconv.Atoi(val[0])
			last, _ := strconv.Atoi(val[1])
			hold = append(hold, first)
			hold = append(hold, last)
		}
		if overlaping(hold) {
			out += 1
		}
	}
	fmt.Println("stage 2 =", out)
}
func fullyHold(in []int) bool {
	first, second := toRange(in)
	if contain(first, in[2]) && contain(first, in[3]) || contain(second, in[0]) && contain(second, in[1]) {
		return true
	}
	return false
}
func overlaping(in []int) bool {
	first, second := toRange(in)
	for _, i := range first {
		if contain(second, i) {
			return true
		}
	}
	return false
}
func toRange(in []int) ([]int, []int) {
	var first []int
	var second []int
	for i := in[0]; i <= in[1]; i++ {
		first = append(first, i)
	}
	for i := in[2]; i <= in[3]; i++ {
		second = append(second, i)
	}
	return first, second
}
func contain(in []int, val int) bool {
	for _, i := range in {
		if i == val {
			return true
		}
	}
	return false
}

File: day4/python/day4.py
import os

os.chdir("..")
data = open(os.getcwd()+"/input.txt", "r").read().split("\n")


def input(data: list[str]):
    return [[[int(k) for k in j.split("-")]
             for j in i.split(",")] for i in data]


def main(data: list[str]):
    data_input = input(data)
    stage1 = 0
    for i in data_input:
        if i[0][0] in range(i[1][0], i[1][1]+1) and i[0][1] in range(i[1][0], i[1][1]+1):
            stage1 += 1
            continue
        if i[1][0] in range(i[0][0], i[0][1]+1) and i[1][1] in range(i[0][0], i[0][1]+1):
            stage1 += 1
    stage2 = len([i for i in data_input if len([j for j in range(i[0][0], i[0][1]+1) if j in [k for k in range(i[1][0], i[1][1]+1)]])
                 != 0])
    print("stage 1 = ", stage1)
    print("stage 2 = ", stage2)


main(data)
